pyccel/ast/headers.py

```diff
@@ -179,6 +179,32 @@ def create_definition(self):
         funcs = []
         dtypes = []
 
+        def build_argument(var_name, dc):
+            #Constructs an argument variable from a dictionary.
+            dtype    = dc['datatype']
+            allocatable = dc['allocatable']
+            is_pointer = dc['is_pointer']
+            precision = dc['precision']
+            rank = dc['rank']
+
+            order = None
+            shape = None
+            if rank >1:
+                order = dc['order']
+
+            if isinstance(dtype, str):
+                try:
+                    dtype = datatype(dtype)
+                except ValueError:
+                    dtype = DataTypeFactory(str(dtype), ("_name"))()
+                    is_pointer = True
```

Comment:
What case is this treating? Why is it always a pointer?

Reply:
it's been there before, i only moved it another function, I don't know why we set `is_pointer` to true. this is the case of a user defined type i guess?

Reply:
Maybe, but I'm not sure why that would make it a pointer (in the Fortran sense). Do you know @saidctb

Reply:
this is the case of a user defined class , it's not used anymore for sure

Reply:
should i remove it and raise an error instead ?

Reply:
I would leave it as we plan to add classes at some point so there's no point re writing what we already have. I'm not sure why a class as an argument would have to be a pointer though

Reply:
in the C case, i think it makes sense, in this example:
```
class p:
    def __init__(self):
        self.x = 0

def f(p):
     p.x = 1

cls = p()
f(cls)
```

in c we can have somthing like this:
```
 typedef struct{
     int x;
     void(*init)(int*);
 } p;

int f(p* cls)
{
   cls->x = 6;
}
```
if instead we used `int f(p cls)` this wont affect the attributes of cls

Reply:
In C lots of things need to be passed as pointers. In your example `cls` should be passed as a pointer anyway in order to avoid costly copies. This is why we created the `stored_in_c_pointer` function.
However things with the `is_pointer` property will me pointers in all languages.
In Fortran (and therefore pyccel as pyccel was originally designed with only Fortran in mind) a pointer is something which points at something else and therefore requires the `pointer` attribute in Fortran. There is no reason for this to be the case for all classes.
You could also imagine the case of C++ where we would use references rather than pointers for your example

Reply:
If we can't think of a reason why this needs to be a pointer I would like to remove this line to avoid potential errors once classes are added. It's not really relevant to this PR, but if we don't do it now then we will forget. Do you agree @saidctb ?

Reply:
yes fine by me

Reply:
Then I will approve once this line is removed

Reply:
@rmahjoubi Your PR is ready to be merged as soon as you apply the requested change!

Reply:
@EmilyBourne It's good now